### PR TITLE
Improve subscribers instrumentation

### PR DESCRIPTION
**Pull request by mgiacomini**
Hi, we are using this library at my current job and it's very nice and helpful, thanks for the work :)

Afterwhile was necessary to send more data from the message that is processed to appsignal, like correlation id's, message body, etc. I had added some functionality based on `Coduit.Message` struct and would be nice if we had this on the lib.

### CODE REVIEW

File: lib/plug/subscriber.ex [hunk 1]
```diff
@@ -20,7 +20,38 @@ defmodule ConduitAppsignal.Plug.Subscriber do
...
     action = "#{inspect(opts[:subscriber])}.process"
-    Appsignal.Transaction.set_action(action)
+
+    tags = %{
+      source: message.source,
+      destination: message.destination,
+      user_id: message.user_id,
+      correlation_id: message.correlation_id,
+      message_id: message.message_id
+    }
```
Comment (blatyo) [inline lib/plug/subscriber.ex:30]: This stuff makes sense

File: lib/plug/subscriber.ex [hunk 2]
```diff
@@ -20,7 +20,38 @@ defmodule ConduitAppsignal.Plug.Subscriber do
...
+
+    environment =
+      Map.merge(
+        %{
+          content_type: message.content_type,
+          content_encoding: message.content_encoding,
+          created_by: message.created_by,
+          created_at: message.created_at
+        },
```
Comment (blatyo) [inline lib/plug/subscriber.ex]: As do these

File: lib/plug/subscriber.ex [hunk 3]
```diff
@@ -20,7 +20,38 @@ defmodule ConduitAppsignal.Plug.Subscriber do
...
+        message.headers
+      )
+
+    custom = %{
+      assigns: message.assigns,
+      private: message.private
```
Comment (blatyo) [inline lib/plug/subscriber.ex]: Things in here may not be serializable. For example, tuples and functions. Also, there may be large data structures or secrets kept in here that people may not want to send to a third party.

File: lib/plug/subscriber.ex [hunk 4]
```diff
@@ -20,7 +20,38 @@ defmodule ConduitAppsignal.Plug.Subscriber do
...
+    }
+    
+    action
+    |> Appsignal.Transaction.set_action()
+    |> Appsignal.Transaction.set_sample_data("environment", environment)
+    |> Appsignal.Transaction.set_sample_data("tags", tags)
+    |> Appsignal.Transaction.set_sample_data("params", message.body)
```
Comment (blatyo) [inline lib/plug/subscriber.ex]: Message bodies may be very large or contain secrets.

File: lib/plug/subscriber.ex [hunk 5]
```diff
@@ -20,7 +20,38 @@ defmodule ConduitAppsignal.Plug.Subscriber do
...
+
+    environment =
+      Map.merge(
+        %{
+          content_type: message.content_type,
+          content_encoding: message.content_encoding,
+          created_by: message.created_by,
+          created_at: message.created_at
+        },
+        message.headers
```
Comment (blatyo) [inline lib/plug/subscriber.ex]: Headers may also contain secrets.

**Comment by blatyo** [comment]
Apologies for not seeing this sooner. I think parts of this make a lot of sense to include by default. Some of it, I think the user should opt into. Like I don't think the body should be sent by default because it could contain private information (e.g. user email) you may not want to leak to a third party service like appsignal. There's also the serialization problem, where some message parts are not guaranteed to be serializable. So, the messages to appsignal would likely silently fail on the machine while trying to send to appsignal. For example, perhaps you've put a struct in assigns, but didn't define a json serializer for it. Finally, sometimes the fields contain a massive amount of information and you don't want to send it to appsignal because of that.

So, I think these fields need to be opted into in some way:
* message.headers
* message.assigns
* message.private
* message.body

**Comment by mgiacomini** [comment]
> information

When I wrote this code I take the Phoenix controllers integration with appsignal as inspiration, which sends the request body and headers. But you are right and I'll remove the *optional* data. 

Maybe another approach to it would be to create new plugs like `ConduitAppsignal.Plug.Body`, `ConduitAppsignal.Plug.Headers` or something like that, WDYT?

Thanks for the reply :D

**Comment by blatyo** [comment]
I like the idea of doing them as separate plugs like you suggest. Seems like a good way to bundle up the functionality and allow people to opt into it.

**Comment by blatyo** [comment]
0.6.3 has these changes in it

**Pull request merged**